qcodes/instrument_drivers/Spectrum/M4i.py

```diff
@@ -777,22 +788,39 @@ def start_acquisition(self, mV_range, memsize, posttrigger_size=None, verbose=0)
 
         return {'memsize': memsize, 'numch': numch, 'mV_range': mV_range}
 
-    def _transfer_buffer_numpy(self, memsize, numch):
-        """ Transfer buffer to numpy array """
+    def _transfer_buffer_numpy(self, memsize: int, numch: int, bytes_per_sample=2) -> np.ndarray:
+        """ Transfer buffer to numpy array
+
+        Args:
+            memsize (int): number of samples to transfer
+            numch (int): number of channels
+            bytes_per_sample (int): specifies the datatype. 2 for int16, 4 for int32
+        Returns:
+            array: transfered data
+
+        """
         # setup software buffer
-        buffer_size = ct.c_int16 * memsize * numch
-        data_buffer = (buffer_size)()
+        sample_ctype: Union[Type[ct.c_int16], Type[ct.c_int32]]
+        if bytes_per_sample == 2:
+            sample_ctype = ct.c_int16
+        elif bytes_per_sample == 4:
+            sample_ctype = ct.c_int32
+        else:
+            raise ValueError('bytes_per_sample should be 2 or 4')
+
+        ctype_buffer_type = sample_ctype * memsize * numch
+        data_buffer = (ctype_buffer_type)()
         data_pointer = ct.cast(data_buffer, ct.c_void_p)
 
         # data acquisition
         self._def_transfer64bit(
-            pyspcm.SPCM_BUF_DATA, pyspcm.SPCM_DIR_CARDTOPC, 0, data_pointer, 0, 2 * memsize * numch)
+            pyspcm.SPCM_BUF_DATA, pyspcm.SPCM_DIR_CARDTOPC, 0, data_pointer, 0, bytes_per_sample * memsize * numch)
         self.general_command(pyspcm.M2CMD_DATA_STARTDMA |
                              pyspcm.M2CMD_DATA_WAITDMA)
 
         # convert buffer to numpy array
-        data = ct.cast(data_pointer, ct.POINTER(buffer_size))
-        output = np.frombuffer(data.contents, dtype=ct.c_int16)
+        data = ct.cast(data_pointer, ct.POINTER(ctype_buffer_type))
+        output = np.frombuffer(data.contents, dtype=sample_ctype)

```

Comment:
isn't it possible to use `data_buffer` here instead of the casting and `data.contents`?

Reply:
@astafan8 Thanks for the suggestion. I updated and tested the code.